### PR TITLE
Cifar100 example

### PR DESCRIPTION
**Pull request by SvecDavid**

### CODE REVIEW

File: cifar100/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+from .cifar_dataset import CIFARDataset
+from .wide_resnet import WideResNet
```
Comment (blazekadam) [inline cifar100/__init__.py]: I would prefer not to include the net here. The reason is that it includes `tensorflow` which takes a while. Therefore every time a dataset method is invoked, one need to wait a few seconds for no actual reason.

File: cifar100/cifar_dataset.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
+import logging
+import tarfile
+import pickle
+import os
+import os.path as path
+import shutil
+
+import numpy as np
+import cxflow as cx
+from matplotlib import pyplot as plt
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: This package is not included in the installation instructions. Actually, we should have a `requirements.txt` file in the project root probably.

File: cifar100/cifar_dataset.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+DOWNLOAD_URL = ['https://github.com/Cognexa/cxflow-examples/releases/download/cifar100-dataset/cifar-100-python.tar.gz']
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: Perhaps you want to rename this to `DOWNLOAD_URLS` as the variable is a list. Alternatively, make it a string and create the list when needed later on.

File: cifar100/cifar_dataset.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+FILENAME = 'cifar-100-python.tar.gz'
+
+DOWNLOAD_STRUCTURE = {'train': {'images': b'data', 'labels': b'fine_labels'},
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: The name `DOWNLOAD_STRUCTURE` is a bit misleading, the data are already downloaded at this point. Perhaps `DATA_STRUCTURE` or `FILE_STRUCTURE` would be a bit better.

File: cifar100/cifar_dataset.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                  'test': {'images': b'data', 'labels': b'fine_labels'},
+                  'meta': {'label_names': b'fine_label_names'}}
+
+
+class CIFARDataset(cx.DownloadableDataset):
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: Consider moving the constants above to this class which is the only consumer of these constants. If need be, one should access it through the class so that their affiliation is stated explicitly.

File: cifar100/cifar_dataset.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    """ Cifar100 dataset for image classification."""
+
+    def _configure_dataset(self, data_root=path.join('cifar100', '.cifar-data'), batch_size:int=100, label_class:int=-1, **kwargs) -> None:
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: Missing docstring for the parameters, especially the meaning and purpose of `label_class` parameter is not clear at this moment.

File: cifar100/cifar_dataset.py (hunk 5 above)
Comment (blazekadam) [reply cifar100/cifar_dataset.py]: \+ missing docstring

File: cifar100/cifar_dataset.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        self._batch_size = batch_size
+        self._data_root = data_root
+        self._download_urls = DOWNLOAD_URL
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: Do we need to copy this variable to `self`? Is there a reason not to use it directly?

File: cifar100/cifar_dataset.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        self._label_class = label_class
+        self._data = {'train': {}, 'test': {}, 'label_names': None}
+        self._data_loaded = False
+        self._mean = None
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: Do we need to store `mean` and `std` in `self`? It is used like a local variable later on.

It may be handy for later inference where the training data are no longer available though. Perhaps you want to save as a yaml to the output_dir (this variable is passed in the constructor **kwargs).

File: cifar100/cifar_dataset.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        self._std = None
+
+    def _load_data(self, normalize: bool=True) -> None:
+        if not self._data_loaded:
+            logging.info('Loading Cifar100 data to memory')
+
+            file_path = path.join(self.data_root, FILENAME)
+            if not path.exists(file_path):
+                raise FileNotFoundError('File `{}` does not exist. '
+                                        'Run `cxflow dataset download <path-to-config>` first!'.format(file_path))
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: I would substitute `<path-to-config>` with `cifar100`.

File: cifar100/cifar_dataset.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+            with tarfile.open(file_path, 'r') as files:
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: We do not need to extract it every time the data are loaded. Consider wrapping the `download` method if need be.

File: cifar100/cifar_dataset.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+                self._data['train']['images'] = self._data['train']['images'].reshape(self._data['train']['images'].shape[0], 3, 32, 32).transpose(0,2,3,1).astype("float32")
+
+                self._mean = np.mean(self._data['train']['images'], axis=0, keepdims=True)
+                self._std = np.std(self._data['train']['images'])
+
+                self._data['train']['images'] = (self._data['train']['images'] - self._mean) / self._std
+                self._data['test']['images'] = (self._data['test']['images'] - self._mean) / self._std
+
+            else:
+                self._data['test']['images'] = self._data['test']['images'].reshape(self._data['test']['images'].shape[0], 3, 32, 32).transpose(0,2,3,1).astype("uint8")
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: This code may is duplicite to lines 55 and 56, it can be placed before this branching.

File: cifar100/cifar_dataset.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+                self._data['train']['images'] = self._data['train']['images'].reshape(self._data['train']['images'].shape[0], 3, 32, 32).transpose(0,2,3,1).astype("uint8")
+
+
+            shutil.rmtree(path.join(self.data_root, 'cifar-100-python'))
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: Why?

File: cifar100/cifar_dataset.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        self._batch_size = batch_size
+        self._data_root = data_root
+        self._download_urls = DOWNLOAD_URL
+        self._label_class = label_class
+        self._data = {'train': {}, 'test': {}, 'label_names': None}
+        self._data_loaded = False
+        self._mean = None
+        self._std = None
+
+    def _load_data(self, normalize: bool=True) -> None:
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: I would expected normalization turned off by default, but feel free to keep it like this.

Comment (blazekadam) [reply cifar100/cifar_dataset.py]: \+ missing docstring!

File: cifar100/cifar_dataset.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+            else:
+                self._data['test']['images'] = self._data['test']['images'].reshape(self._data['test']['images'].shape[0], 3, 32, 32).transpose(0,2,3,1).astype("uint8")
+                self._data['train']['images'] = self._data['train']['images'].reshape(self._data['train']['images'].shape[0], 3, 32, 32).transpose(0,2,3,1).astype("uint8")
+
+
+            shutil.rmtree(path.join(self.data_root, 'cifar-100-python'))
+
+            self._data_loaded = True
+
+    def grid_of_images(self) -> None:
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: missing docstring

File: cifar100/cifar_dataset.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        self._load_data(normalize=False)
+
+        if self._label_class >= 0:
+            indexes = [i for i, label in enumerate(self._data['train']['labels']) if label == self._label_class][:100]
+            i = -1
+
+        for col in range(10):
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: This is not very *pythonic* and *numpy styled* way of doing things. 
You should be able to do all this with 3 or 4 lines of more readable code.

- prepare the index array (either range or range mapped through `indexes`)
- map the indices to images
- use `numpy` ops to create a single array (`vstack`, `hstack`, `reshape` etc.)

File: cifar100/cifar_dataset.py [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+                else:
+                    row_array = self._data['train']['images'][index]
+
+            if col != 0:
+                output_array = np.concatenate((output_array, row_array), axis=0)
+            else:
+                output_array = row_array
+            row_array = None
+
+        plt.axis('off')
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: You are using a plotting library and functions yet you do not plot at all. If you want to save an image, use some image library such as opencv.

File: cifar100/cifar_dataset.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+        plt.imsave('cifar100_images.png', output_array)
+
+    def train_stream(self) -> cx.Stream:
+        self._load_data()
```
Comment (blazekadam) [inline cifar100/cifar_dataset.py]: You should shuffle data in the train stream every iteration.

File: cifar100/config.yaml [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+hooks:
+  - cxflow_tensorflow.hooks.DecayLR:
+      decay_value: 0.98
+      variable_name: learning_rate
+      decay_type: multiply
+  - ComputeStats:
+      variables: [loss, accuracy]
+  - LogVariables
+  - SaveBest:
+      stream: test
```
Comment (blazekadam) [inline cifar100/config.yaml]: You should not use the test data for model selection!!

File: cifar100/readme.md [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+3. (Optional) Test your loaded images. You have two options:
+  1. Save grid of first 100 images from loaded train images.
+  ```
+  cxflow dataset grid_of_images cifar100
+  ```
+  This will save new image in your cxflow-examples folder.
+
+  2. Save grid of first 100 images from specific class. You have to specify which class do you want to save in your config file. Just set in dataset block which class and again call:
+  ```
+  label_class: 0
```
Comment (blazekadam) [inline cifar100/readme.md]: This line is not an command line. It should not be here. You may call the **cxflow** like this though
```
cxflow dataset grid_of_images cifar100 dataset.label_class=0
```

File: cifar100/cifar_dataset.py (hunk 14 above)
Comment (blazekadam) [reply cifar100/cifar_dataset.py]: I would rename the method to `visualize`.

File: cifar100/readme.md [hunk 20]
```diff
@@ -0,0 +1,44 @@
+# CIFAR-100 Dataset
```
Comment (blazekadam) [inline cifar100/readme.md]: *# Wide residual network with CIFAR-100 Dataset*

File: cifar100/wide_resnet.py [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+class WideResNet(cxtf.BaseModel):
+
+    def _create_model(self, depth: int, k: int, weight_decay: float=0.0005, dropout: float=0, num_classes: int=100):
+
+        images = tf.placeholder(tf.float32, shape=[None, 32, 32, 3], name='images')
+        labels = tf.placeholder(tf.int64, shape=[None], name='labels')
+
+        net = wrn_model(images, depth, k, weight_decay, dropout, num_classes, is_training=True)
+
+        loss = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels, logits=net)
```
Comment (blazekadam) [inline cifar100/wide_resnet.py]: Can you name the tensor right here? Not sure...

File: cifar100/wide_resnet.py [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+        for layer, filters in enumerate(num_filters[1:]):
+            net = wide_block(net, filters, n, layer+2, dropout, is_training)
+
+        net = slim.avg_pool2d(net, 8, stride=1, padding='VALID')
+        net = slim.flatten(net)
+        net = slim.fully_connected(net, num_classes, activation_fn=None)
+
+    return net
+
+class WideResNet(cxtf.BaseModel):
```
Comment (blazekadam) [inline cifar100/wide_resnet.py]: Missing docstring (with reference).

File: cifar100/wide_resnet.py [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+
+    def _create_model(self, depth: int, k: int, weight_decay: float=0.0005, dropout: float=0, num_classes: int=100):
```
Comment (blazekadam) [inline cifar100/wide_resnet.py]: Missing docstring! What these params suppose to do?

**Comment by blazekadam** [comment]
I do miss CI tests here.

**Comment by jablkoj** [comment]
Stale Pull Request, closing.
